### DISCUSSION TITLE
How to get pylance to realize the underscore function `_` in `gettext` module?

### DISCUSSION THREAD

**Discussion by davuses** [microsoft/pyright#5670] [Q&A]
I'm working with `gettext` module which assigns method `gettext.gettext()` to underscore `_` in the following way:

```python
import builtins
builtins.__dict__['_'] = self.gettext
```

After calling: `gettext.translation().install()`, name `_` can be accessed as a method, but Pylance gives me this error: `"_" is not define`.\
Any ideas how to handle this?

**Comment by erictraut** [comment]
The `gettext` module is dynamically monkey-patching Python's global namespace. I can't stress enough how bad this is. I would personally avoid any library that is doing such ill-advised hackery. There's no excuse for it. It should export its symbols like every other library.

That said, pyright has a feature that can help you out. You can read about it [here](https://microsoft.github.io/pyright/#/builtins).

**Reply by davuses** [reply]
**Update**: 
The trick I provided here doesn't work because the built-in name `"_"` `gettext` relies upon will be overridden by the function `_`  , an alternative way could be:
```py
translation.install()

_ = builtins.__dict__["_"]
```
 ---
**Old:**
I totally agree with you that this is an ugly and error-prone hacking which should be avoided. 

The [Extending Builtins](https://microsoft.github.io/pyright/#/builtins?id=extending-builtins) solution you suggest is insufficient for me, because I have other linters like pylint and ruff on and it cannot handle their complaints. I come up with one trick that could stop the complaints from all the linters, which is to define a function named `_` before calling the `install()` method and the dynamic patching:
```py
def _(text: str):
    return text

translation.install()

print(_("No monkey-patching!"))
```
 Unfortunately this trick hurts readability, I hope there could be a better idea.

---

**Reply by erictraut** [reply]
I can't offer you a solution for other linters. You would need to consult the maintainers of those tools to see if they have solutions to work around this hacky behavior. The solution I suggested should work fine for pyright and pylance.

I think the ideal solution is to convince the maintainer of this library to abandon this hacky behavior. You could then import the symbol using an `import` statement like every other Python library: `from gettext import _` or better yet something like `from gettext import translate`.

**Reply by davuses** [reply]
I'm not sure if this could be done considering `gettext` module is part of the standard library. Anyway, thanks for your help and suggestions!
